File: src/main/java/com/wesker926/leetcode/algorithms/p0869/s3/Solution.java
package com.wesker926.leetcode.algorithms.p0869.s3;

import java.util.HashSet;
import java.util.Set;

/**
 * @author wesker.gh
 * @date 2021/10/28
 * @description count and compare
 */
public class Solution {

    private final Set<String> set;

    {
        set = new HashSet<>();
        for (int i = 1; i <= 1000000000; i <<= 1) {
            set.add(numToStr(i));
        }
    }

    public boolean reorderedPowerOf2(int n) {
        return set.contains(numToStr(n));
    }

    private String numToStr(int n) {
        char[] cs = new char[10];
        for (; n != 0; n /= 10) {
            cs[n % 10]++;
        }
        return new String(cs);
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0869/s3/solution.go
// Package s3
// @author: wesker.gh
// @date: 2021/10/28
// @description: count and compare
package s3

var cache = initCache()

func initCache() map[string]interface{} {
	c, n := make(map[string]interface{}), 1
	for ; n <= 1000000000; n <<= 1 {
		c[numToStr(n)] = nil
	}
	return c
}

func reorderedPowerOf2(n int) bool {
	_, ok := cache[numToStr(n)]
	return ok
}

func numToStr(n int) string {
	cs := make([]byte, 10)
	for ; n != 0; n /= 10 {
		cs[n%10]++
	}
	return string(cs)
}
